[Gradient Compression] Add an index field to GradBucket for PowerSGD

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48757 [Gradient Compression] Add an index field to GradBucket for PowerSGD**

Add an index field to GradBucekt, so error_dict is keyed by this index instead of the hashcode of input tensor.

Howevever, sometimes the buckets can be rebuilt in the forward pass. In this case, the shape of the bucket with the same index will not be consistent with the one in the previous iteration, and hence the error tensor will be re--initialized as a zero tensor of the new shape.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25288496](https://our.internmc.facebook.com/intern/diff/D25288496/)

Commit:
[Gradient Compression] Add an index field to GradBucket for PowerSGD

Add an index field to GradBucekt, so error_dict is keyed by this index instead of the hashcode of input tensor.

Howevever, sometimes the buckets can be rebuilt in the forward pass. In this case, the shape of the bucket with the same index will not be consistent with the one in the previous iteration, and hence the error tensor will be re--initialized as a zero tensor of the new shape.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25288496](https://our.internmc.facebook.com/intern/diff/D25288496/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add an index field to GradBucket for PowerSGD"


Add an index field to GradBucekt, so error_dict is keyed by this index instead of the hashcode of input tensor.

Howevever, sometimes the buckets can be rebuilt in the forward pass. In this case, the shape of the bucket with the same index will not be consistent with the one in the previous iteration, and hence the error tensor will be re--initialized as a zero tensor of the new shape.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25288496](https://our.internmc.facebook.com/intern/diff/D25288496/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add an index field to GradBucket for PowerSGD"


Add an index field to GradBucekt, so error_dict is keyed by this index instead of the hashcode of input tensor.

Howevever, sometimes the buckets can be rebuilt in the forward pass. In this case, the shape of the bucket with the same index will not be consistent with the one in the previous iteration, and hence the error tensor will be re--initialized as a zero tensor of the new shape.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25288496](https://our.internmc.facebook.com/intern/diff/D25288496/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1558,7 +1565,9 @@ static const auto DistributedC10dFrontendTorchBind =
           return c10d_frontend_singleton;
         }))
-        .def("new_process_group_helper", &::c10d::DistributedC10d::newProcessGroupHelper);
+        .def(
```
Nit: It would be great to have formatting-only changes in a separate diff if they are required.

SciPioneer replied:
Sure. There are not too many such changes in this commit. Will try to avoid such case if too many such lines are involved.

rohan-varma reviewed (approved):
LGTM, please ensure CI is green

Commit:
Update on "[Gradient Compression] Add an index field to GradBucket for PowerSGD"


Add an index field to GradBucekt, so error_dict is keyed by this index instead of the hashcode of input tensor.

Howevever, sometimes the buckets can be rebuilt in the forward pass. In this case, the shape of the bucket with the same index will not be consistent with the one in the previous iteration, and hence the error tensor will be re--initialized as a zero tensor of the new shape.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25288496](https://our.internmc.facebook.com/intern/diff/D25288496/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add an index field to GradBucket for PowerSGD"


Add an index field to GradBucekt, so error_dict is keyed by this index instead of the hashcode of input tensor.

Howevever, sometimes the buckets can be rebuilt in the forward pass. In this case, the shape of the bucket with the same index will not be consistent with the one in the previous iteration, and hence the error tensor will be re--initialized as a zero tensor of the new shape.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25288496](https://our.internmc.facebook.com/intern/diff/D25288496/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add an index field to GradBucket for PowerSGD"


Add an index field to GradBucekt, so error_dict is keyed by this index instead of the hashcode of input tensor.

Howevever, sometimes the buckets can be rebuilt in the forward pass. In this case, the shape of the bucket with the same index will not be consistent with the one in the previous iteration, and hence the error tensor will be re--initialized as a zero tensor of the new shape.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25288496](https://our.internmc.facebook.com/intern/diff/D25288496/)

[ghstack-poisoned]
